### PR TITLE
Retrieve archives that have logs in the given search time window

### PR DESCRIPTION
**Pull request by abvarun226**
# Description
When using MySQL as metadata db, only retrieve archive directories that have logs in the given search time window. This will improve search performance.

# Validation performed
Functional tests on Ubuntu bionic.

### CODE REVIEW

**Comment by Yadumathur** [comment]
@kirkrodrigues / @jackluo923  As discussed yesterday can you please review the PR and let us know if you further see any changes in PR or its okay to be merged ?
As provided description this is to improve the search performance.

**Comment by Yadumathur** [comment]
ref : https://github.com/y-scope/clp/issues/139

File: components/core/src/GlobalMySQLMetadataDB.hpp
Comment (kirkrodrigues) [inline components/core/src/GlobalMySQLMetadataDB.hpp:69]: Instead of modifying the existing iterator, could we create a new one for filtering by timestamps? E.g., `get_archive_iterator_for_time_range`

File: components/core/src/GlobalMySQLMetadataDB.cpp
Comment (kirkrodrigues) [inline components/core/src/GlobalMySQLMetadataDB.cpp:201]: A few comments:
1. If I'm not mistaken, I think this condition will miss some relevant archives since it's looking for files (and their corresponding archives) that are completely *within* the time range, rather than *overlapping* with the time range. E.g., if the user is searching for a results in the time range `[5, 10]`, this SQL won't find an archive containing files which have the time range `[4, 6]`. I think ideally we'd make the condition `file_begin_ts <= {user_end_ts} AND file_end_ts >= {user_begin_ts}`.
2. We should be using `>=` and `<=` since the time range from `clg` is inclusive.
3. The DISTINCT and JOIN ops may become expensive when a user has a lot of files. Instead, I think we could compute the begin and end timestamp of an archive at compression time and store it directly in the archives table. I can create a branch demonstrating this if it sounds good to you.

File: components/core/src/clg/clg.cpp
Comment (kirkrodrigues) [inline components/core/src/clg/clg.cpp:95]: Assuming we create a separate iterator just for filtering by time, then we should add a branch here for handling the case where the user specifies a time range.

File: components/core/src/clp/decompression.cpp
Comment (kirkrodrigues) [inline components/core/src/clp/decompression.cpp:73]: Assuming we create a separate iterator for filtering by time, then we shouldn't need to modify this file.

File: components/core/src/GlobalSQLiteMetadataDB.hpp
Comment (kirkrodrigues) [inline components/core/src/GlobalSQLiteMetadataDB.hpp:74]: Do you see any concerns with implementing the time range filtering for SQLite databases?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Thanks for contributing! I've raised a few concerns, but overall, the idea sounds good.

File: components/core/src/GlobalMySQLMetadataDB.cpp
Comment (Yadumathur) [inline components/core/src/GlobalMySQLMetadataDB.cpp:201]:
```suggestion
    auto statement_string = fmt::format("SELECT DISTINCT {}{}.{} FROM {}{} JOIN {}{} ON {}{}.{} = {}{}.{} WHERE {}{}.{} >= {} AND {}{}.{} <= {} ORDER BY {} ASC, {} ASC",
```

Comment (Yadumathur) [reply components/core/src/GlobalMySQLMetadataDB.cpp:201]: Added suggestive insert to line , I think that is what you are expecting ?
Point 3 : Please provide the feature details and branch , we would be happy to look at that.

Can you also add myself and @abvarun226 in contributors to the CLP repo?

File: components/core/src/GlobalMetadataDB.hpp
Comment (Yadumathur) [inline components/core/src/GlobalMetadataDB.hpp:75]:
```suggestion
    virtual ArchiveIterator* get_archive_iterator () = 0;
/**
     * Gets an iterator to iterate over every archive that contains a given start and end time stamp in the global metadata database
     * @return The archive iterator
     */
virtual ArchiveIterator* get_archive_iterator_for_time_range (epochtime_t begin_ts, epochtime_t end_ts) = 0;
```

File: components/core/src/GlobalMySQLMetadataDB.cpp
Comment (abvarun226) [reply components/core/src/GlobalMySQLMetadataDB.cpp:201]: @kirkrodrigues
1. That makes sense, I've updated the condition in SQL statement to `file_begin_ts <= {user_end_ts} AND file_end_ts >= {user_begin_ts}`
2. I've updated to use inclusive conditions
3. Looks like this will require changes to archives table schema, I'd be happy to incorporate that here. As of current schema, I see files table also has archive_id, just that we will not be able to sort by creation idx

Comment (abvarun226) [reply components/core/src/GlobalMySQLMetadataDB.cpp:201]: will need more changes than just `>=` and `<=`. Updated it in my commit

File: components/core/src/clg/clg.cpp
Comment (abvarun226) [reply components/core/src/clg/clg.cpp:95]: I've added a branch where I check if the begin_ts and end_ts is supplied by the user.

File: components/core/src/GlobalMySQLMetadataDB.hpp
Comment (abvarun226) [reply components/core/src/GlobalMySQLMetadataDB.hpp]: I've added a new iterator `get_archive_iterator_for_time_window`

File: components/core/src/clp/decompression.cpp
Comment (abvarun226) [reply components/core/src/clp/decompression.cpp]: Yes, we don't need this. I've reverted this change

File: components/core/src/GlobalSQLiteMetadataDB.hpp
Comment (abvarun226) [reply components/core/src/GlobalSQLiteMetadataDB.hpp]: No concerns really, I've added the implementation for SQLite database as well.

File: components/core/src/GlobalMySQLMetadataDB.cpp
Comment (kirkrodrigues) [reply components/core/src/GlobalMySQLMetadataDB.cpp]: 1 & 2 look good, thanks! Yeah, 3 requires:
* Adding the begin and end timestamp to the archive schema
* Tracking and updating the begin and end timestamp for the archive in streaming_archive::writer::Archive (probably in `close_segment_and_persist_file_metadata`).
* Changing `GlobalMetadataDB::update_archive_size` to also update the begin and end timestamp.

I'll let you know when I have a branch with those changes so you can take a look or if you feel like you have a rough idea of how to do it, feel free to incorporate it.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Looks good except for one missed change, and hopefully we can get the archive metadata change in before merging.

File: components/core/src/GlobalMySQLMetadataDB.cpp
Comment (kirkrodrigues) [reply components/core/src/GlobalMySQLMetadataDB.cpp]: > Can you also add myself and @abvarun226 in contributors to the CLP repo?

Sorry, forgot to reply to this. Once the PR is merged, GitHub will automatically add anyone who has commits in this PR to the contributors list.

Comment (kirkrodrigues) [reply components/core/src/GlobalMySQLMetadataDB.cpp]: Sorry for the wait, here's the commit: https://github.com/kirkrodrigues/clp/commit/4fd12eed6e60613f11d99cf2b0e5bc95cbe87007

If it works for you, I could push it directly to your PR or you can.

**Review by kirkrodrigues — APPROVED** [review]

**Comment by kirkrodrigues** [comment]
Thanks for your contribution!

**Pull request merged**
